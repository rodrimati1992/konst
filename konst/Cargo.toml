[package]
name = "konst"
version = "0.3.5"
authors = ["rodrimati1992 <rodrimatt1985@gmail.com>"]
rust-version = "1.65.0"
edition = "2021"
license = "Zlib"
description = "Const equivalents of std functions, compile-time comparison, and parsing"
documentation = "https://docs.rs/konst/"
readme="../README.md"
keywords = ["no-std", "const_fn", "parsing"]
categories = ["no-std", "parsing"]
repository = "https://github.com/rodrimati1992/konst/"
include = [
    "Cargo.toml", 
    "src/**/*.rs", 
    "../README.md",
    "LICENSE-ZLIB.md", 
]


[dependencies.konst_proc_macros]
version = "=0.3.0"
path = "../konst_proc_macros"
optional = true

[dependencies.konst_kernel]
version = "=0.3.5"
path = "../konst_kernel"
features = ["rust_1_64", "__for_konst"]

[dev-dependencies.rand]
version = "0.8.4"
default-features = false
features = ["small_rng"]

# dev-dependencies can't be optional
[dependencies.trybuild]
version = "1.0"
optional = true

[dependencies.typewit]
version = "1.1"
features = ["rust_1_61"]

[dependencies.const_panic]
version = "0.2.6"
features = ["rust_1_64"]
default-features = false

[features]
default = ["cmp", "iter", "parsing_proc"]


# Enables extra checks for debug
debug = ["konst_kernel/debug"]
cmp = []
iter = ["konst_kernel/iter"]
parsing_proc = ["parsing", "konst_proc_macros"]
parsing = []
rust_latest_stable = []
mut_refs = ["rust_latest_stable", "konst_kernel/mut_refs"]
nightly_mut_refs = ["mut_refs", "konst_kernel/nightly_mut_refs"]

alloc = ["konst_kernel/alloc"]

# Enables ui tests, which are intended for the latest stable release.
# This is not enabled in CI, because it's way too sensitive to 
# changes in how rustc formats errors
__ui = ["trybuild", "rust_latest_stable"]
__cp_derive = ["const_panic/derive", "const_panic/non_basic"]

docsrs = ["konst_kernel/docsrs"]



[package.metadata.docs.rs]
features = ["docsrs", "alloc", "cmp", "parsing_proc", "rust_latest_stable", "nightly_mut_refs"]

