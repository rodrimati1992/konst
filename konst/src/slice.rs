//! `const fn` equivalents of slice methods.
//!
//! # Removed in 0.3.0
//!
//! These functions were removed in 0.3.0 because there is an equivalent
//! const fn in the standard library:
//!
//! - `first`
//! - `last`
//! - `split_first`
//! - `split_last`
//!
//!
//!

/// `const fn`s for comparing slices for equality and ordering.
#[cfg(feature = "cmp")]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "cmp")))]
pub mod cmp;

mod bytes_pattern;
mod slice_as_chunks;
mod slice_concatenation;
mod slice_const_methods;

#[cfg(feature = "iter")]
mod slice_iter_methods;

pub use bytes_pattern::BytesPattern;

pub(crate) use bytes_pattern::PatternNorm;

pub use self::slice_as_chunks::*;
pub use self::slice_concatenation::*;
pub use self::slice_const_methods::*;

#[cfg(feature = "iter")]
pub use slice_iter_methods::*;

__declare_slice_cmp_fns! {
    import_path = "konst",

    (
        ///
        ///  # Example
        ///
        ,
        /// ```rust
        /// use konst::slice::eq_bytes;
        ///
        /// const FOO: &[u8] = b"foo";
        /// const BAR: &[u8] = b"fooooo";
        /// const BAZ: &[u8] = b"bar";
        ///
        ///
        /// const FOO_EQ_FOO: bool = eq_bytes(FOO, FOO);
        /// assert!( FOO_EQ_FOO );
        ///
        /// const FOO_EQ_BAR: bool = eq_bytes(FOO, BAR);
        /// assert!( !FOO_EQ_BAR );
        ///
        /// const FOO_EQ_BAZ: bool = eq_bytes(FOO, BAZ);
        /// assert!( !FOO_EQ_BAZ );
        ///
        /// ```
        ///
        ,
        /// ```rust
        /// use konst::slice::cmp_bytes;
        ///
        /// use std::cmp::Ordering;
        ///
        /// const FOO: &[u8] = b"foo";
        /// const BAR: &[u8] = b"fooooo";
        /// const BAZ: &[u8] = b"bar";
        ///
        ///
        /// const FOO_CMP_FOO: Ordering = cmp_bytes(FOO, FOO);
        /// assert_eq!(FOO_CMP_FOO, Ordering::Equal);
        ///
        /// const FOO_CMP_BAR: Ordering = cmp_bytes(FOO, BAR);
        /// assert_eq!(FOO_CMP_BAR, Ordering::Less);
        ///
        /// const FOO_CMP_BAZ: Ordering = cmp_bytes(FOO, BAZ);
        /// assert_eq!(FOO_CMP_BAZ, Ordering::Greater);
        ///
        /// ```
        ///
        ,
        u8,
        eq_bytes,
        cmp_bytes,
    )
}

__declare_fns_with_docs! {
    (Option<&'a [u8]>, (eq_option_bytes, cmp_option_bytes))

    docs(default)

    macro = __impl_option_cmp_fns!(
        for['a,]
        params(l, r)
        eq_comparison = eq_bytes(l, r),
        cmp_comparison = cmp_bytes(l, r),
        parameter_copyability = copy,
    ),
}

/// The error produced by trying to convert from
/// `&[T]` to `&[T; N]`, or from `&mut [T]` to `&mut [T; N]`.
#[doc(inline)]
pub use konst_kernel::slice::slice_for_konst::TryIntoArrayError;

/// Tries to convert from `&[T]` to `&[T; N]`.
///
/// Returns an `Err(TryIntoArrayError{..})` when the slice doesn't match the expected length.
///
/// # Example
///
/// ```rust
/// use konst::{
///     slice::{TryIntoArrayError, try_into_array},
///     result,
///     unwrap_ctx,
/// };
///
///
/// const fn arr_5() -> Option<&'static [u64; 5]> {
///     let slice: &[u64] = &[1, 10, 100, 1000, 10000];
///
///     // Passing the length explicitly to the function
///     result::ok!(try_into_array::<_, 5>(slice))
/// }
///
/// assert_eq!(arr_5(), Some(&[1, 10, 100, 1000, 10000]));
///
///
/// const fn err() -> Result<&'static [u64; 5], TryIntoArrayError> {
///     let slice: &[u64] = &[];
///
///     // Letting the function infer the length of the array,
///     try_into_array(slice)
/// }
///
/// assert!(err().is_err());
///
///
/// const fn arr_3() -> &'static [u64; 3] {
///     let slice: &[u64] = &[3, 5, 8];
///
///     let array = unwrap_ctx!(try_into_array(slice));
///     
///     // You can destructure the array into its elements like this
///     let [a, b, c] = *array;
///     
///     array
/// }
///
/// assert_eq!(arr_3(), &[3, 5, 8]);
///
/// ```
///
/// [`try_into_array`]: ./macro.try_into_array.html
#[doc(inline)]
pub use konst_kernel::slice::slice_for_konst::try_into_array_func as try_into_array;

/// Tries to convert from `&mut [T]` to `&mut [T; N]`.
///
/// Returns an `Err(TryIntoArrayError{..})` when the slice doesn't match the expected length.
///
/// # Example
///
/// ```rust
/// # #![feature(const_mut_refs)]
/// use konst::{slice, unwrap_ctx};
///
/// const fn mut_array_from<const LEN: usize>(slice: &mut [u8], from: usize) -> &mut [u8; LEN] {
///     let sliced = slice::slice_range_mut(slice, from, from + LEN);
///     unwrap_ctx!(slice::try_into_array_mut(sliced))
/// }
///
/// # fn main() {
///
/// let slice = &mut [3, 5, 8, 13, 21, 34, 55, 89, 144, 233];
///
/// let foo: &mut [u8; 2] = mut_array_from(slice, 0);
/// assert_eq!(foo, &mut [3, 5]);
///
/// let bar: &mut [u8; 3] = mut_array_from(slice, 2);
/// assert_eq!(bar, &mut [8, 13, 21]);
///
/// let baz: &mut [u8; 4] = mut_array_from(slice, 4);
/// assert_eq!(baz, &mut [21, 34, 55, 89]);
///
/// # }
/// ```
///
#[cfg(feature = "mut_refs")]
#[cfg_attr(
    feature = "docsrs",
    doc(cfg(any(feature = "mut_refs", feature = "nightly_mut_refs")))
)]
#[doc(inline)]
pub use konst_kernel::slice::slice_for_konst::try_into_array_mut_func as try_into_array_mut;
