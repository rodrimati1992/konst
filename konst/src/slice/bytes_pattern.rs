use crate::{
    chr,
    polymorphism::{HasTypeWitness, MakeTypeWitness, TypeEq, TypeWitnessTypeArg},
};

/// A byte slice pattern.
///
/// Types that implement this trait can be used to search into a byte slice.
///
/// This trait can only be implemented in the `konst` crate.
///
pub trait BytesPattern<const N: usize>: HasTypeWitness<BytesPatternInput<N, Self>> {
    #[doc(hidden)]
    type __Normalized: ?Sized;
}

#[non_exhaustive]
pub enum BytesPatternInput<const N: usize, P: ?Sized + BytesPattern<N>> {
    Str(TypeEq<P, str>, TypeEq<*const P::__Normalized, *const [u8]>),
    Bytes(TypeEq<P, [u8]>, TypeEq<*const P::__Normalized, *const [u8]>),
    Array(
        TypeEq<P, [u8; N]>,
        TypeEq<*const P::__Normalized, *const [u8]>,
    ),
    Char(
        TypeEq<P, char>,
        TypeEq<*const P::__Normalized, *const chr::Utf8Encoded>,
    ),
}

impl<const N: usize, Arg> TypeWitnessTypeArg for BytesPatternInput<N, Arg>
where
    Arg: ?Sized + BytesPattern<N>,
{
    type Arg = Arg;
}

impl MakeTypeWitness for BytesPatternInput<0, str> {
    const MAKE: Self = BytesPatternInput::Str(TypeEq::NEW, TypeEq::NEW);
}
impl BytesPattern<0> for str {
    #[doc(hidden)]
    type __Normalized = [u8];
}

impl MakeTypeWitness for BytesPatternInput<0, [u8]> {
    const MAKE: Self = BytesPatternInput::Bytes(TypeEq::NEW, TypeEq::NEW);
}
impl BytesPattern<0> for [u8] {
    #[doc(hidden)]
    type __Normalized = [u8];
}

impl<const N: usize> MakeTypeWitness for BytesPatternInput<N, [u8; N]> {
    const MAKE: Self = BytesPatternInput::Array(TypeEq::NEW, TypeEq::NEW);
}
impl<const N: usize> BytesPattern<N> for [u8; N] {
    #[doc(hidden)]
    type __Normalized = [u8];
}

impl MakeTypeWitness for BytesPatternInput<0, char> {
    const MAKE: Self = BytesPatternInput::Char(TypeEq::NEW, TypeEq::NEW);
}
impl BytesPattern<0> for char {
    #[doc(hidden)]
    type __Normalized = chr::Utf8Encoded;
}

#[derive(Copy, Clone)]
pub(crate) enum PatternNorm<'a, const N: usize, P>
where
    P: ?Sized + BytesPattern<N>,
{
    Bytes {
        val: &'a [u8],
        ten: TypeEq<*const P::__Normalized, *const [u8]>,
    },
    Char {
        val: chr::Utf8Encoded,
        ten: TypeEq<*const P::__Normalized, *const chr::Utf8Encoded>,
    },
}

impl<'a, const N: usize, P> PatternNorm<'a, N, P>
where
    P: ?Sized + BytesPattern<N>,
{
    pub(crate) const fn new(pattern: &'a P) -> Self {
        match P::WITNESS {
            BytesPatternInput::Str(te, ten) => {
                let val: &'a [u8] = te.in_ref().to_right(pattern).as_bytes();
                PatternNorm::Bytes { val, ten }
            }
            BytesPatternInput::Bytes(te, ten) => {
                let val: &'a [u8] = te.in_ref().to_right(pattern);
                PatternNorm::Bytes { val, ten }
            }
            BytesPatternInput::Array(te, ten) => {
                let val: &'a [u8] = te.in_ref().to_right(pattern);
                PatternNorm::Bytes { val, ten }
            }
            BytesPatternInput::Char(te, ten) => {
                let val: &'a char = te.in_ref().to_right(pattern);
                let val = chr::encode_utf8(*val);
                PatternNorm::Char { val, ten }
            }
        }
    }

    pub(crate) const fn as_bytes(&self) -> &[u8] {
        match self {
            PatternNorm::Bytes { val, ten } => ten.reachability_hint(val),
            PatternNorm::Char { val, ten } => ten.reachability_hint(val.as_bytes()),
        }
    }
}
