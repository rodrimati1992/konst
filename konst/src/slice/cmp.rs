#[doc(inline)]
pub use crate::slice::{
    cmp_bytes as cmp_slice_u8, cmp_option_bytes as cmp_option_slice_u8, eq_bytes as eq_slice_u8,
    eq_option_bytes as eq_option_slice_u8,
};

use core::cmp::Ordering;

__declare_slice_cmp_fns! {
    import_path = "konst",

    (,,, u16, eq_slice_u16, cmp_slice_u16,)
    (,,, u32, eq_slice_u32, cmp_slice_u32,)
    (,,, u64, eq_slice_u64, cmp_slice_u64,)
    (,,, u128, eq_slice_u128, cmp_slice_u128,)
    (,,, usize, eq_slice_usize, cmp_slice_usize,)

    (,,, i8, eq_slice_i8, cmp_slice_i8,)
    (,,, i16, eq_slice_i16, cmp_slice_i16,)
    (,,, i32, eq_slice_i32, cmp_slice_i32,)
    (,,, i64, eq_slice_i64, cmp_slice_i64,)
    (,,, i128, eq_slice_i128, cmp_slice_i128,)
    (,,, isize, eq_slice_isize, cmp_slice_isize,)

    (,,, bool, eq_slice_bool, cmp_slice_bool,)
    (,,, char, eq_slice_char, cmp_slice_char,)
}

__delegate_const_eq! {
    skip_coerce;

    /// Compares two `&[&str]` for equality.
    pub const fn eq_slice_str(ref l: &[&str], r: &[&str]) -> bool {
        crate::const_eq_for!(slice; l, r, crate::eq_str)
    }
}

__delegate_const_ord! {
    skip_coerce;

    /// Compares two `&[&str]`, returning the order of `left` relative to `right`.
    pub const fn cmp_slice_str(ref left: &[&str], right: &[&str]) -> Ordering {
        crate::const_cmp_for!(slice; left, right, crate::cmp_str)
    }
}

__delegate_const_eq! {
    skip_coerce;

    /// Compares two `&[&[u8]]` for equality.
    pub const fn eq_slice_bytes(ref l: &[&[u8]], r: &[&[u8]]) -> bool {
        crate::const_eq_for!(slice; l, r, eq_slice_u8)
    }
}

__delegate_const_ord! {
    skip_coerce;

    /// Compares two `&[&[u8]]`, returning the order of `left` relative to `right`.
    pub const fn cmp_slice_bytes(ref left: &[&[u8]], right: &[&[u8]]) -> Ordering {
        crate::const_cmp_for!(slice; left, right, cmp_slice_u8)
    }
}

__declare_fns_with_docs! {
    (Option<&'a [u16]>, (eq_option_slice_u16, cmp_option_slice_u16))
    (Option<&'a [u32]>, (eq_option_slice_u32, cmp_option_slice_u32))
    (Option<&'a [u64]>, (eq_option_slice_u64, cmp_option_slice_u64))
    (Option<&'a [u128]>, (eq_option_slice_u128, cmp_option_slice_u128))
    (Option<&'a [usize]>, (eq_option_slice_usize, cmp_option_slice_usize))
    (Option<&'a [i8]>, (eq_option_slice_i8, cmp_option_slice_i8))
    (Option<&'a [i16]>, (eq_option_slice_i16, cmp_option_slice_i16))
    (Option<&'a [i32]>, (eq_option_slice_i32, cmp_option_slice_i32))
    (Option<&'a [i64]>, (eq_option_slice_i64, cmp_option_slice_i64))
    (Option<&'a [i128]>, (eq_option_slice_i128, cmp_option_slice_i128))
    (Option<&'a [isize]>, (eq_option_slice_isize, cmp_option_slice_isize))
    (Option<&'a [bool]>, (eq_option_slice_bool, cmp_option_slice_bool))
    (Option<&'a [char]>, (eq_option_slice_char, cmp_option_slice_char))

    docs(default)

    macro = __impl_option_cmp_fns!(
        for['a,]
        params(l, r)
        eq_comparison = crate::cmp::CmpWrapper(l).const_eq(r),
        cmp_comparison = crate::cmp::CmpWrapper(l).const_cmp(r),
        parameter_copyability = copy,
    ),
}

__declare_fns_with_docs! {
    (Option<&'a [&'b str]>, (eq_option_slice_str, cmp_option_slice_str))
    (Option<&'a [&'b [u8]]>, (eq_option_slice_bytes, cmp_option_slice_bytes))

    docs(default)

    macro = __impl_option_cmp_fns!(
        for['a, 'b,]
        params(l, r)
        eq_comparison = crate::cmp::CmpWrapper(l).const_eq(r),
        cmp_comparison = crate::cmp::CmpWrapper(l).const_cmp(r),
        parameter_copyability = copy,
    ),
}
