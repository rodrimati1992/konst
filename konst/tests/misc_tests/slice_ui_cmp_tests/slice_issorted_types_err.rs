use konst::slice;

struct NoCmp;

const _: bool = slice::is_sorted!([NoCmp]);
const _: () = slice::is_sorted!([0u8]);

const _: bool = slice::is_sorted_by!([0u8; 0], |x, y| -> u8 { 0u8 });
const _: bool = slice::is_sorted_by!([0u8; 0], |x, y| 0u8);
const _: bool = slice::is_sorted_by!([0u8; 0], |x: &u16, y| false);
const _: bool = slice::is_sorted_by!([0u8; 0], |x, y: &u16| false);
const _: () = slice::is_sorted_by!([0u8; 0], |x, y| -> bool { false });

const _: bool = slice::is_sorted_by_key!([0u8; 0], |x| NoCmp);
const _: bool = slice::is_sorted_by_key!([0u8; 0], |_: &u16| 0u8);
const _: () = slice::is_sorted_by_key!([0u8; 0], |x| x);




fn main() {}