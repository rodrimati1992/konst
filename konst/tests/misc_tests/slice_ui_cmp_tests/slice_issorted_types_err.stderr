error[E0277]: the trait bound `NoCmp: ConstCmp` is not satisfied
 --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:5:17
  |
5 | const _: bool = slice::is_sorted!([NoCmp]);
  |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `ConstCmp` is not implemented for `NoCmp`
  |
  = help: the following other types implement trait `ConstCmp`:
            &T
            &mut T
            CmpWrapper<P>
            NonZero<i128>
            NonZero<i16>
            NonZero<i32>
            NonZero<i64>
            NonZero<i8>
          and $N others
  = note: required for `&NoCmp` to implement `ConstCmp`
note: required by a bound in `__AssertConstCmp`
 --> src/cmp/const_cmp.rs
  |
  | pub struct __AssertConstCmp<'a, T: ConstCmp> {
  |                                    ^^^^^^^^ required by this bound in `__AssertConstCmp`
  = note: this error originates in the macro `$crate::cmp::const_cmp` which comes from the expansion of the macro `slice::is_sorted` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `NoCmp: ConstCmp` is not satisfied
 --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:5:17
  |
5 | const _: bool = slice::is_sorted!([NoCmp]);
  |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `ConstCmp` is not implemented for `NoCmp`
  |
  = help: the following other types implement trait `ConstCmp`:
            &T
            &mut T
            CmpWrapper<P>
            NonZero<i128>
            NonZero<i16>
            NonZero<i32>
            NonZero<i64>
            NonZero<i8>
          and $N others
  = note: required for `&NoCmp` to implement `ConstCmp`
note: required by a bound in `__AssertConstCmp`
 --> src/cmp/const_cmp.rs
  |
  | pub struct __AssertConstCmp<'a, T: ConstCmp> {
  |                                    ^^^^^^^^ required by this bound in `__AssertConstCmp`
  = note: this error originates in the macro `$crate::cmp::const_cmp` which comes from the expansion of the macro `slice::is_sorted` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `NoCmp: ConstCmp` is not satisfied
 --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:5:17
  |
5 | const _: bool = slice::is_sorted!([NoCmp]);
  |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `ConstCmp` is not implemented for `NoCmp`
  |
  = help: the following other types implement trait `ConstCmp`:
            &T
            &mut T
            CmpWrapper<P>
            NonZero<i128>
            NonZero<i16>
            NonZero<i32>
            NonZero<i64>
            NonZero<i8>
          and $N others
  = note: required for `&NoCmp` to implement `ConstCmp`
note: required by a bound in `IsAConstCmp::<<R as ConstCmp>::Kind, <R as ConstCmp>::This, R>::NEW`
 --> src/cmp/const_cmp.rs
  |
  |     R: ?Sized + ConstCmp,
  |                 ^^^^^^^^ required by this bound in `IsAConstCmp::<<R as ConstCmp>::Kind, <R as ConstCmp>::This, R>::NEW`
...
  |     pub const NEW: Self = Self(PhantomData);
  |               --- required by a bound in this associated constant
  = note: this error originates in the macro `$crate::__coerce_to_cmp2` which comes from the expansion of the macro `slice::is_sorted` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `NoCmp: ConstCmp` is not satisfied
 --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:5:17
  |
5 | const _: bool = slice::is_sorted!([NoCmp]);
  |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `ConstCmp` is not implemented for `NoCmp`
  |
  = help: the following other types implement trait `ConstCmp`:
            &T
            &mut T
            CmpWrapper<P>
            NonZero<i128>
            NonZero<i16>
            NonZero<i32>
            NonZero<i64>
            NonZero<i8>
          and $N others
  = note: required for `&NoCmp` to implement `ConstCmp`
note: required by a bound in `IsAConstCmp::<<R as ConstCmp>::Kind, <R as ConstCmp>::This, R>::NEW`
 --> src/cmp/const_cmp.rs
  |
  |     R: ?Sized + ConstCmp,
  |                 ^^^^^^^^ required by this bound in `IsAConstCmp::<<R as ConstCmp>::Kind, <R as ConstCmp>::This, R>::NEW`
...
  |     pub const NEW: Self = Self(PhantomData);
  |               --- required by a bound in this associated constant
  = note: this error originates in the macro `$crate::__coerce_to_cmp2` which comes from the expansion of the macro `slice::is_sorted` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0282]: type annotations needed
 --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:5:17
  |
5 | const _: bool = slice::is_sorted!([NoCmp]);
  |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type
  |
  = note: this error originates in the macro `$crate::cmp::const_cmp` which comes from the expansion of the macro `slice::is_sorted` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
 --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:6:15
  |
6 | const _: () = slice::is_sorted!([0u8]);
  |               ^^^^^^^^^^^^^^^^^^^^^^^^ expected `()`, found `bool`
  |
  = note: this error originates in the macro `$crate::__slice_is_sorted_impl` which comes from the expansion of the macro `slice::is_sorted` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
 --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:8:17
  |
8 | const _: bool = slice::is_sorted_by!([0u8; 0], |x, y| -> u8 { 0u8 });
  |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `bool`, found `u8`
  |
  = note: this error originates in the macro `$crate::__eval_closure` which comes from the expansion of the macro `slice::is_sorted_by` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
 --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:9:17
  |
9 | const _: bool = slice::is_sorted_by!([0u8; 0], |x, y| 0u8);
  |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `bool`, found `u8`
  |
  = note: this error originates in the macro `$crate::__eval_closure` which comes from the expansion of the macro `slice::is_sorted_by` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:10:17
   |
10 | const _: bool = slice::is_sorted_by!([0u8; 0], |x: &u16, y| false);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&u16`, found `&u8`
   |
   = note: expected reference `&u16`
              found reference `&u8`
   = note: this error originates in the macro `slice::is_sorted_by` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:11:17
   |
11 | const _: bool = slice::is_sorted_by!([0u8; 0], |x, y: &u16| false);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&u16`, found `&u8`
   |
   = note: expected reference `&u16`
              found reference `&u8`
   = note: this error originates in the macro `slice::is_sorted_by` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:12:15
   |
12 | const _: () = slice::is_sorted_by!([0u8; 0], |x, y| -> bool { false });
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `()`, found `bool`
   |
   = note: this error originates in the macro `$crate::__slice_is_sorted_impl` which comes from the expansion of the macro `slice::is_sorted_by` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `NoCmp: ConstCmp` is not satisfied
  --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:14:17
   |
14 | const _: bool = slice::is_sorted_by_key!([0u8; 0], |x| NoCmp);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `ConstCmp` is not implemented for `NoCmp`
   |
   = help: the following other types implement trait `ConstCmp`:
             &T
             &mut T
             CmpWrapper<P>
             NonZero<i128>
             NonZero<i16>
             NonZero<i32>
             NonZero<i64>
             NonZero<i8>
           and $N others
note: required by a bound in `__AssertConstCmp`
  --> src/cmp/const_cmp.rs
   |
   | pub struct __AssertConstCmp<'a, T: ConstCmp> {
   |                                    ^^^^^^^^ required by this bound in `__AssertConstCmp`
   = note: this error originates in the macro `$crate::cmp::const_cmp` which comes from the expansion of the macro `slice::is_sorted_by_key` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `NoCmp: ConstCmp` is not satisfied
  --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:14:17
   |
14 | const _: bool = slice::is_sorted_by_key!([0u8; 0], |x| NoCmp);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `ConstCmp` is not implemented for `NoCmp`
   |
   = help: the following other types implement trait `ConstCmp`:
             &T
             &mut T
             CmpWrapper<P>
             NonZero<i128>
             NonZero<i16>
             NonZero<i32>
             NonZero<i64>
             NonZero<i8>
           and $N others
note: required by a bound in `__AssertConstCmp`
  --> src/cmp/const_cmp.rs
   |
   | pub struct __AssertConstCmp<'a, T: ConstCmp> {
   |                                    ^^^^^^^^ required by this bound in `__AssertConstCmp`
   = note: this error originates in the macro `$crate::cmp::const_cmp` which comes from the expansion of the macro `slice::is_sorted_by_key` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `NoCmp: ConstCmp` is not satisfied
  --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:14:17
   |
14 | const _: bool = slice::is_sorted_by_key!([0u8; 0], |x| NoCmp);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `ConstCmp` is not implemented for `NoCmp`
   |
   = help: the following other types implement trait `ConstCmp`:
             &T
             &mut T
             CmpWrapper<P>
             NonZero<i128>
             NonZero<i16>
             NonZero<i32>
             NonZero<i64>
             NonZero<i8>
           and $N others
note: required by a bound in `IsAConstCmp::<<R as ConstCmp>::Kind, <R as ConstCmp>::This, R>::NEW`
  --> src/cmp/const_cmp.rs
   |
   |     R: ?Sized + ConstCmp,
   |                 ^^^^^^^^ required by this bound in `IsAConstCmp::<<R as ConstCmp>::Kind, <R as ConstCmp>::This, R>::NEW`
...
   |     pub const NEW: Self = Self(PhantomData);
   |               --- required by a bound in this associated constant
   = note: this error originates in the macro `$crate::__coerce_to_cmp2` which comes from the expansion of the macro `slice::is_sorted_by_key` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `NoCmp: ConstCmp` is not satisfied
  --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:14:17
   |
14 | const _: bool = slice::is_sorted_by_key!([0u8; 0], |x| NoCmp);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `ConstCmp` is not implemented for `NoCmp`
   |
   = help: the following other types implement trait `ConstCmp`:
             &T
             &mut T
             CmpWrapper<P>
             NonZero<i128>
             NonZero<i16>
             NonZero<i32>
             NonZero<i64>
             NonZero<i8>
           and $N others
note: required by a bound in `IsAConstCmp::<<R as ConstCmp>::Kind, <R as ConstCmp>::This, R>::NEW`
  --> src/cmp/const_cmp.rs
   |
   |     R: ?Sized + ConstCmp,
   |                 ^^^^^^^^ required by this bound in `IsAConstCmp::<<R as ConstCmp>::Kind, <R as ConstCmp>::This, R>::NEW`
...
   |     pub const NEW: Self = Self(PhantomData);
   |               --- required by a bound in this associated constant
   = note: this error originates in the macro `$crate::__coerce_to_cmp2` which comes from the expansion of the macro `slice::is_sorted_by_key` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0282]: type annotations needed
  --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:14:17
   |
14 | const _: bool = slice::is_sorted_by_key!([0u8; 0], |x| NoCmp);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type
   |
   = note: this error originates in the macro `$crate::cmp::const_cmp` which comes from the expansion of the macro `slice::is_sorted_by_key` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:15:17
   |
15 | const _: bool = slice::is_sorted_by_key!([0u8; 0], |_: &u16| 0u8);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^----^^^^^^
   |                 |                                      |
   |                 |                                      expected due to this
   |                 expected `&u16`, found `&u8`
   |
   = note: expected reference `&u16`
              found reference `&u8`
   = note: this error originates in the macro `slice::is_sorted_by_key` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> tests/misc_tests/slice_ui_cmp_tests/slice_issorted_types_err.rs:16:15
   |
16 | const _: () = slice::is_sorted_by_key!([0u8; 0], |x| x);
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `()`, found `bool`
   |
   = note: this error originates in the macro `$crate::__slice_is_sorted_impl` which comes from the expansion of the macro `slice::is_sorted_by_key` (in Nightly builds, run with -Z macro-backtrace for more info)
